### ISSUE TITLE
Issues with DelegateClass.

### ISSUE THREAD

**Issue by tomasmiguez** [ohler55/oj#919]
We started using `DelegateClass` on our codebase recently, and are experiencing some compatibility issues with Oj.
The main problem has already been reported here https://github.com/ohler55/oj/issues/477, but the proposed solution does not work for our use case.
We currently use this options on the `dump` method calls

```ruby
      options = {
        create_additions: false,
        mode: :compat,
        time_format: :xmlschema,
        second_precision: 0,
        use_to_hash: true
      }
```

but if we change it to compat, we have 2 problems:
- `time_format` is not valid for that mode, so we can't keep this change transparent.
- Decimal numbers are getting serialized as strings. We thought this might be fixed by setting `compat_bigdecimal` to false or false, but that did not work either.

How can we get delegate objects to serialize only the underlying object? And not the attributes of the delegate that are anecdotic?

For aditional context, Sequel uses Delegate classes in their implementation and this is causing our problems here: https://github.com/jeremyevans/sequel/discussions/2153

Thanks in advance,
Tomas.

**Comment by ohler55** [comment]
Can you provide more information about what you are looking for. I got the basics but am not sure if the app is a rails app or something else. Also, what is the preferred mode? It sounds like you are not using rails or else the :compat time format would be okay.

If you can use some other mode that does not try to be compatible with the JSON gem there are a few more options to deal with a custom delegate marshalling.

**Comment by ohler55** [comment]
Would it be possible to monkey-patch SimpleDelegate to add a `as_json()` or `to_hash()` method? If so then that method could just call the delegate.

**Comment by ohler55** [comment]
I would advise against just returning `__getobj__` because as you pointed out it is not a hash. `as_json` would be better but again. as you pointed out monkey patching is inherently dangerous and especially so on core classes. We are hemmed in by the language and the classes though. Let me see if I can break down the issue a bit more. What you want is something in Oj that when it sees an object that has a class that inherits from Delegate it runs some special code. Right now Oj supports special or odd encoding based on class but not based on inherited class. Brainstorming it kind of sounds like implementing a parallel class inheritance with associated methods is what you are looking for. I'm not sure to implement that right now. It is an interesting train of thought to follow but non-trivial.

**Comment by tomasmiguez** [comment]
Yea, that's exactly it. 

I did not follow you on the bit about the parallel class inheritance, but looking at the code and if you were open to make modifications to Oj to address this issue, I was thinking that maybe an option that when set checks for `Delegate` with `rb_obj_is_kind_of` around [here](https://github.com/ohler55/oj/blob/develop/ext/oj/custom.c#L824) would do the trick. But I guess that might not be a generic enough solution, and I have no clue about the possible performance implications of such a change.

**Comment by ohler55** [comment]
If every call to dump an object had to check for whether the instance was a type of delegate that would be expensive. If a more generic solution was offered it would be worse. I do not want to make a special case for Delegates but a low performance overhead option might be to provide a hook that is called for each object. If the developer was willing to sacrifice performance they could define hook for what ever classes they wanted.

**Comment by tomasmiguez** [comment]
Yes, that makes sense. The check would have been behind the option, so the cost of it would only have been payed by those interested in this behaviour. But I can see how a hook to modify the object can be a more generic solution for these issue.

Although, now that I think about it, this can be done outside of Oj as well, like before serialization you could ask if the object is a delegate and in that case serialize the internal object instead of the delegate.

**Comment by ohler55** [comment]
It could be done outside Oj but the before hook/code would have to be able to traverse all object attributes to look for delegates and then create a duplicate tree of objects. It will be expensive unless you have a very specific case where delegates are alway top level objects.

**Comment by tomasmiguez** [comment]
Yea you are right, for our current use case we only need to check the top level object, but that won't help for a general solution.

Would you be open to a PR submission exploring the hooks idea from where we could iterate a solution?

**Comment by ohler55** [comment]
As long as it covers the general case, that would be great as long as you don't mind me being a bit picky.
